### PR TITLE
[CODAP-14] Sort categories by default; import category order from v2 documents

### PR DESCRIPTION
**Pull request by kswenson**
[[CODAP-14](https://concord-consortium.atlassian.net/browse/CODAP-14)]

This PR accomplishes two fairly simple (in theory) tasks, one of which required a deceptively complex implementation:
1. Categories are sorted alphanumerically by default. This is accomplished with a handful of lines in `CategorySet`.
2. The category order is imported from v2 documents. A priori, this seems like it should be straightforward, but v3 handles category order differently from v2 and so a non-trivial transformation is required.

V2 serializes category order by simply serializing the entire array of category values in order. To adopt this representation internally and to support serialization and undo/redo, etc., v3 would have to maintain a serializable MST property which listed all of the attribute values in order and was kept up to date with any changes to the data. For many attributes this would essentially mean maintaining a duplicate copy of the data in a different order. But the only useful information in the order is the set of changes the user has explicitly made to the default order and most attributes/category sets are never modified by the user. Therefore, v3 adopts a representation that just tracks the specific changes made by the user. Until the user makes a change, there's nothing to serialize and only those changes explicitly made by the user ever require serialization. Undo/redo of a category order change works automatically with the addition of a single move to the array of user moves. This works great and simplifies the implementation tremendously in all situations _except_ importing from v2 documents, which requires transforming a raw list of ordered strings to a reasonable set of moves that transforms an initial order (sorted alphanumerically) to another order (imported).

There are simple algorithms that work, e.g. move the first string into position, move the second string into position, etc. until done. But if asked to transform `["e", "a", "b", "c", "d"]` to `["a", "b", "c", "d", "e"]`, this will use four moves where one would suffice. This could easily result in representations that were substantially less efficient than a redundant copy of the data. After multiple unsuccessful attempts to get ChatGPT to help at various points in the process I ended up with an approach that is fairly simple to describe, but was tricky to implement. An extensive set of jest tests were used to track down various edge cases which led to the final implementation of the `MinimalMovesFinder` class.

The basic approach is to:
1. Scan the two arrays to identify blocks of strings in correct order as well as the longest such block.
2. If the user has not changed the order then this will be a single block and we're done.
3. While the longest block is not the first block or is the first block and shouldn't be
  a. Identify a block to move and merge it with another block with which it should be contiguous
  b. Merge any merge-able adjacent blocks now that the previous block move occurred.
4. While the longest block is not the last block or is the last block and shouldn't be
  a. Identify a block to move and merge it with another block with which it should be contiguous
  b. Merge any merge-able adjacent blocks now that the previous block move occurred.

The end result should be a single block that contains all of the strings in the correct order, and the set of moves required to get there was accumulated along the way. As always, the devil is in the details. Internally, the implementation represents the blocks in an AVL tree, which is a form of binary search tree. The algorithm requires frequently navigating the tree to find which blocks belong where and so a binary search tree is more efficient than a simple array. The key of each node is the current index of its first string and the values of each node are the strings contained in it. As each move is made, the keys of the affected blocks must be updated to accommodate the change, generally incrementing/decrementing by one. The code also makes use of a `MappedStringArray` class for quickly mapping between strings and indices.

In the worst case scenario this could be terribly inefficient, but the worst case is highly unlikely to occur. The only way for the user to change the order of the categories in v2 is to manually drag them one at a time in the graph. I sincerely doubt there are any documents out there with thousands of categories where the user has painstakingly dragged each of them to its unique position.

Because `CategorySet`s are stored in the `SharedCaseMetadata`, as previously mentioned in an architecture meeting, I modified the `SharedCaseMetadata` to support a provisional `DataSet` so that it can be created/initialized outside of the document tree and then added to the document later. This required updating some jest tests and one method in the `BaseDocumentContent`, but had the salutary effect of eliminating some console errors being thrown by some jest tests.

Finally, since I was in the neighborhood this PR also imports the legacy `decimals` property from v2 documents.

[CODAP-14]: https://concord-consortium.atlassian.net/browse/CODAP-14?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Review by tealefristoe — APPROVED** [review]
I definitely didn't understand all of this, but it looks good to me 👍 

I left a few small suggestions and questions which you can take or leave.

There's probably some subtleties that I'm missing, but if I were to develop a system for representing category order, my first instinct would be to have an optional ordered list of all categories if any of them have been moved, but to use the default order if the list is missing. In the vast majority of cases, no categories have moved, so there would be no list. In the few cases where there is a list, it might be a bit overkill when only one or two categories have been shifted, but I think it would be a lot easier to understand when new engineers need to come in and tinker around in this space.

File: v3/src/models/data/minimal-moves-finder.ts [hunk 1]
```diff
@@ -0,0 +1,490 @@
...
+  }
+
+  // merge one block at the beginning of another block
+  mergeBlockBefore(blockToMoveKey: number, blockToJoinKey: number) {
+    const blockToMoveStrings = this.blockTree.search(blockToMoveKey) ?? []
+    let successorValue = this.currStrings.getValue(blockToJoinKey)
+    if (!successorValue) return
+    for (let i = blockToMoveStrings.length - 1; i >= 0; --i) {
+      const value = blockToMoveStrings[i]
+      ;[blockToMoveKey, blockToJoinKey] = this.moveValueFromBlockToBlock(value, blockToMoveKey, blockToJoinKey, true)
```
[the remark this answers is not in the document]
Comment (kswenson) [reply v3/src/models/data/minimal-moves-finder.ts:359]: When not using explicit semicolons, lines that begin with `[` is one of the places that JavaScript's automatic semicolon insertion (ASI) will _not_ insert one, so we have to put one in explicitly for the code to parse correctly.

File: v3/src/utilities/mapped-string-array.ts [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+    const dstIndex = indexToMove < indexToMoveBefore ? indexToMoveBefore - 1 : indexToMoveBefore
+    return {
+      fromIndex: indexToMove,
+      toIndex: dstIndex,
+      length: this.values.length,
+      before: valueToMoveBefore,
+      after: valueToMoveAfter
+    }
+  }
+}
```
[the remark this answers is not in the document]
Comment (kswenson) [reply v3/src/utilities/mapped-string-array.ts:87]: I played around with this a little bit and the differences between the two are finicky enough that it didn't seem to be a huge improvement so I ended up deciding not to worry about it for now.

File: v3/src/models/data/v2-category-set-importer.ts [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+import { ICategoryMove, ICategorySetSnapshot } from "./category-set"
+import { kellyColors } from "../../utilities/color-utils"
+import { colord } from "colord"
+import { MinimalMovesFinder } from "./minimal-moves-finder"
+
+export type V2CategorySetInput = CodapV2ColorMap | ICodapV2CategoryMap
+
+export function importV2CategorySet(attribute: IAttribute, input: V2CategorySetInput): Maybe<ICategorySetSnapshot> {
+  let moves: ICategoryMove[] = []
+  const colors: Record<string, string> = {}
```
[the remark this answers is not in the document]
Comment (kswenson) [reply v3/src/models/data/v2-category-set-importer.ts:14]: Not really, but I added a comment.

**Comment by kswenson** [comment]
> There's probably some subtleties that I'm missing, but if I were to develop a system for representing category order, my first instinct would be to have an optional ordered list of all categories if any of them have been moved, but to use the default order if the list is missing. In the vast majority of cases, no categories have moved, so there would be no list. In the few cases where there is a list, it might be a bit overkill when only one or two categories have been shifted, but I think it would be a lot easier to understand when new engineers need to come in and tinker around in this space.

You'd still have to either maintain that array in serializable MST storage and keep it up to date through any change/insertion/removal of data values, _or_ maintain it in volatile storage (e.g. a MST `view`) and then special-case serialization and undo/redo. This approach at least limits the complexity to the import process rather than spreading it to other aspects of v3's category-related code.
